linked_list/linked_list.py

```diff
@@ -1,79 +1,180 @@
     def get_at_index(self, index):
-        pass
+        if not self.head:
+            return None
+        
+        i = 0
+        current_node = self.head
+        while current_node: 
+            if i == index:
+                return current_node.value
+            current_node = current_node.next 
+            i += 1 
+
+        return None
 
     # method that returns the value of the last node in the linked list
     # returns None if the linked list is empty
-    # Time Complexity: ?
-    # Space Complexity: ?
+    # Time Complexity: O()
+    # Space Complexity: O(1)
     def get_last(self):
-        pass
+        #if my list is empty
+        if not self.head:
+            return None
+        
+        #if my list is one node 
+        if self.head == self.tail:
+            return self.head.value
+
+        #if my list is longer than just one node
+        return self.tail.value
 
+        
     # method that inserts a given value as a new last node in the linked list
-    # Time Complexity: ?
-    # Space Complexity: ?
+    # Time Complexity: O(1)
+    # Space Complexity: O(1)
     def add_last(self, value):
-        pass
+        if not self.head:
+            self.head = self.tail = Node(value)
+        else:
+            new_node = Node(value, None, self.tail)
+            self.tail.next = new_node
+            self.tail = new_node
 
     # method to return the max value in the linked list
     # returns the data value and not the node
     def find_max(self):
-        pass
+        if not self.head:
+            return None
+
+        max = self.head.value
+        current_node = self.head
+
+        while current_node:
+            if current_node.value > max:
+                max = current_node.value
+            else:
+                current_node = current_node.next
+        
+        return max
 
     # method to delete the first node found with specified value
-    # Time Complexity: ?
-    # Space Complexity: ?
+    # Time Complexity: O(n)
+    # Space Complexity: O(n)
     def delete(self, value):
-        pass
-
+        current_node = self.head
+        while current_node:
+            if current_node.value == value:
+                if current_node == self.head:
+                    self.head = current_node.next
+                    self.head.previous = None
+                    return
+                
+                if current_node == self.tail:
+                    self.tail = current_node.previous
+                    self.tail.next = None
+                    return
+
+                current_node.previous.next = current_node.next
+                current_node.next.previous = current_node.previous
+                return
+            else:
+                current_node = current_node.next
+        
     # method to print all the values in the linked list
-    # Time Complexity: ?
-    # Space Complexity: ?
+    # Time Complexity: O(n)
+    # Space Complexity: O(1) .. or n?
     def visit(self):

```

Comment:
✨ Space complexity would be O(n) here since we are creating a new `helper_list` which end up being the length of our linked list